python/ray/train/_internal/checkpoint_manager.py

```diff
@@ -0,0 +1,218 @@
+import heapq
+import logging
+import numbers
+from typing import Any, Dict, List, Optional, Set, Tuple
+
+from ray._private.dict import flatten_dict
+from ray.air.config import MAX
+from ray.air._internal.util import is_nan
+from ray.train import CheckpointConfig
+from ray.train._internal.storage import _delete_fs_path
+from ray.train.checkpoint import Checkpoint
+
+
+logger = logging.getLogger(__name__)
+
+
+class _TrackedCheckpoint:
+    """Checkpoint tracked by a checkpoint manager."""
+
+    def __init__(
+        self, checkpoint: Checkpoint, metrics: Dict[str, Any], index: int = -1
+    ):
+        self.checkpoint = checkpoint
+        self.metrics = metrics
+        self.index = index
+
+
+class _HeapCheckpointWrapper:
+    def __init__(self, priority: Any, tracked_checkpoint: _TrackedCheckpoint):
+        # Ex: 2 wrapped checkpoints, with checkpoint_score_order = MAX
+        # Priority of ckpt 1: (not_is_nan=True, metric=5, id=0)
+        # Priority of ckpt 2: (not_is_nan=True, metric=6, id=1)
+        # In this case, the min-heap should pop ckpt 1 first, since
+        # the metric is smaller.
+        self.priority = priority
+        self.tracked_checkpoint = tracked_checkpoint
+
+    def __lt__(self, other: "_HeapCheckpointWrapper") -> bool:
+        return self.priority < other.priority
+
+    def __repr__(self) -> str:
+        return (
+            f"_HeapCheckpointWrapper(\n"
+            "  priority={self.priority}, "
+            f"tracked_checkpoint={repr(self.tracked_checkpoint)}\n)"
+        )
+
+
+class _CheckpointManager:
```

Comment:
I am going to update this PR with the checkpoint manager usage in a bit, can wait for that before a full review. TLDR is that it will be used in place of the current `Trial.checkpoint_manager`.